internal/terraform/node_resource_validate.go

```diff
@@ -383,10 +384,31 @@ func (n *NodeValidatableResource) validateResource(ctx EvalContext) tfdiags.Diag
 				moreDiags := schema.StaticValidateTraversal(traversal)
 				diags = diags.Append(moreDiags)
 
-				// TODO: we want to notify users that they can't use
-				// ignore_changes for computed attributes, but we don't have an
-				// easy way to correlate the config value, schema and
-				// traversal together.
+				// ignore_changes cannot be used for Computed attributes,
+				// unless they are also Optional.
+				// If the traversal was valid, convert it to a cty.Path and
+				// use that to check whether the Attribute is Computed and
+				// non-Optional.
+				if !diags.HasErrors() {
+					path := traversalToPath(traversal)
+
+					attrSchema := schema.AttributeByPath(path)
+
+					if attrSchema != nil && !attrSchema.Optional && attrSchema.Computed {
+						// ignore_changes uses absolute traversal syntax in config despite
+						// using relative traversals, so we strip the leading "." added by
+						// FormatCtyPath for a better error message.
+						attrDisplayPath := strings.TrimPrefix(tfdiags.FormatCtyPath(path), ".")
+
+						diags = diags.Append(&hcl.Diagnostic{
+							Severity: hcl.DiagWarning,
+							Summary:  "Invalid ignore_changes",
+							Detail:   fmt.Sprintf("ignore_changes cannot be used with non-Optional Computed attributes. Please remove \"%s\" from ignore_changes.", attrDisplayPath),
```

Comment:
This error message includes some provider-developer-facing jargon, but it seems like the audience of the message is module authors who aren't expected to be familiar with that jargon.

I must admit I'm not sure _how_ exactly to describe this particular subset of attributes in user-facing terminology, though. So far we've avoided actually naming this set by using error messages like "This attribute cannot be set in the configuration" -- describing the behavior directly, rather than the name of the behavior -- but I'm not seeing a _great_ way to follow that example here without making a pretty clunky sentence. Here's an _attempt_, at least...

> The ignore_changes argument tells Terraform to ignore future changes to a configured argument after the object has been created, retaining the originally-configured value.
>
> The attribute `%s` is decided by the provider alone and therefore there can be no configured value to compare with. Including this attribute in ignore_changes has no effect. Remove the attribute from ignore_changes to quiet this warning.

My two main goals with this revision are to firstly include some context about what `ignore_changes` is, in case the person who sees this is not the person who originally wrote the module and needs some reassurance that the thing we're pointing at is indeed redundant, and then secondly to try to use a more direct description of the situation rather than relying on the provider protocol schema field names. We also don't typically say "please" in messages emitted from Terraform, but of course that's just a subjective style thing and not the main point of my suggested revisions here.

I might also suggest making the summary be "Redundant ignore_changes element", rather than "Invalid". It's a subtle thing, but I guess it feels a little better to me to report this as "this does nothing" rather than "this is wrong", given that we're presumably going to begin showing this for existing modules that may be maintained by someone other than the user who is seeing the message and I would rather not prompt that downstream user to misunderstand this as having been a breaking change in the module itself, and thus potentially sour the relationship between the user and the module author unnecessarily.